library/modules/EventManager.cpp

```diff
@@ -211,7 +211,12 @@ static int32_t lastJobId = -1;
 static unordered_set<int32_t> startedJobs;
 
 //job completed
-static unordered_map<int32_t, df::job*> prevJobs;
+struct JobCompleteData {
+    int32_t completion_timer;
+    uint32_t flags_bits_repeat;
+};
+static std::unordered_map<int32_t, df::job*> savedJobs;
+static std::unordered_map<int32_t, JobCompleteData> prevJobs;
```

Comment:
I wonder if `prevJobs` should actually be a vector rather than a map, extending `JobCompleteData` with the job id. 

If I understand correctly, DF keeps the job list ordered by job id, so this would automatically ensure that `nowJobs` and `prevJobs` are ordered as well. Thus, when iterating over one and locating data in the other, lookup can be done in *total* time O(n) without hashing, by keeping the index where we last found (or didn't find) the  associated data.

Reply:
I'm sure it's doable for some slight gains, and possibly bigger than slight though -- only way to know is to try. 

It's more complicated due to repeat jobs retaining IDs and there are times when there's large holes in the job id from previous generation.

Reply:
I don't think large holes actually matter. Sure, when there is a large hole, the index into the second vector may remain stationary for a while. Thus, to be fully precise, I suppose the complexity is O(n1+n2) equality checks. One either advances to the next `nowJob` or to the next `prevJob`.

And repeat jobs need to be checked, just as before. This is only a data structure change, the algorithm remains the same.